### ISSUE TITLE
':compileKotlin' is not up-to-date when reusing the configuration-cache

### ISSUE THREAD

**Issue by CristianGM** [gradle/gradle#19266]
### Expected Behavior

Running `./gradlew compileKotlin` a second time without changes I would expect it to reuse the configuration-cache and to have all the tasks (just that one) up-to-date

### Current Behavior

The second time, the configuration is reused but the task is not up-to-date
> Task ':compileKotlin' is not up-to-date because: Value of input property 'filteredArgumentsMap' has changed for task ':compileKotlin' The input changes require a full rebuild for incremental task ':compileKotlin'.

Printing `filteredArgumentsMap` value, type and hash has shown that it's always the same. That's very confusing.

### Context

reusing the configuration shouldn't change the results, also, it means it's making the build slower reducing the benefit of the configuration-cache

### Steps to Reproduce 

https://github.com/CristianGM/filteredArgumentsMap-issue

### Your Environment

Gradle 7.3
Kotlin 1.6.0

**Comment by Petrakeas** [comment]
I'm facing a similar issue. How can I print `filteredArgumentsMap` ?

**Comment by gmiklos-ltg** [comment]
Even without enabling configuration cache this seemingly also breaks Kotlin incremental compilation for me due to `filteredArgumentsMap` always being marked as changed:
```
2022-01-25T12:13:48.685+0100 [DEBUG] [org.gradle.api.Task] [KOTLIN] [IC] Use module detection: true
2022-01-25T12:13:48.687+0100 [DEBUG] [org.gradle.api.Task] [KOTLIN] [IC] Non-incremental compilation will be performed: UNKNOWN_CHANGES_IN_GRADLE_INPUTS
2022-01-25T12:13:48.687+0100 [DEBUG] [org.gradle.api.Task] [KOTLIN] [IC] Clearing output on rebuild
```

**Comment by eskatos** [comment]
Possibly related to https://youtrack.jetbrains.com/issue/KT-49915

**Comment by mlopatkin** [comment]
This seems to be fixed in Gradle 8.0-milestone-5, but I don't have a clue what exactly fixed it. Bisect isn't possible because nightlies used to build Gradle between 8.0-M4 and 8.0-M5 are now gone.

**Comment by bilal2141** [comment]
:app:compileDevDebugKotlin
This task frequently determines build duration because of dependencies between its inputs/outputs and other tasks.  

Duration: 198.5s / 57.1% 
Sub-project: :app 
Plugin: kotlin-android 
Type: org.jetbrains.kotlin.gradle.tasks.KotlinCompile 
Task Execution Categories: Kotlin, Compilation  

Warnings 
No warnings found  

Reason task ran 
`Value of input property 'filteredArgumentsMap' has changed for task ':app:compileDevDebugKotlin' 
``Value of input property 'moduleName' has changed for task ':app:compileDevDebugKotlin' 
``Value of input property 'ownModuleName' has changed for task ':app:compileDevDebugKotlin'`


This task always runs, is there any way to detect why filteredArgumentsMap is marked as changed.

I am on Gradle version 8.1.1
